src/main/java/org/opensearch/security/dlic/rest/api/AuthFailureListenersApiAction.java

```diff
@@ -0,0 +1,217 @@
+
+                        return allowedKeys.put(TYPE_JSON_PROPERTY, DataType.STRING)
+                            .put(AUTHENTICATION_BACKEND_JSON_PROPERTY, DataType.STRING)
+                            .put(ALLOWED_TRIES_JSON_PROPERTY, DataType.INTEGER)
+                            .put(TIME_WINDOW_SECONDS_JSON_PROPERTY, DataType.INTEGER)
+                            .put(BLOCK_EXPIRY_JSON_PROPERTY, DataType.INTEGER)
+                            .put(MAX_BLOCKED_CLIENTS_JSON_PROPERTY, DataType.INTEGER)
+                            .put(MAX_TRACKED_CLIENTS_JSON_PROPERTY, DataType.INTEGER)
+                            .build();
+                    }
+                });
+            }
+        };
+    }
+
+    private ToXContent authFailureContent(final ConfigV7 config) {
+        return (builder, params) -> {
+            builder.startObject();
+            for (String name : config.dynamic.auth_failure_listeners.getListeners().keySet()) {
+                ConfigV7.AuthFailureListener listener = config.dynamic.auth_failure_listeners.getListeners().get(name);
+                builder.startObject(name);
+                builder.field(NAME_JSON_PROPERTY, name)
+                    .field(TYPE_JSON_PROPERTY, listener.type)
+                    .field(AUTHENTICATION_BACKEND_JSON_PROPERTY, listener.authentication_backend)
+                    .field(ALLOWED_TRIES_JSON_PROPERTY, listener.allowed_tries)
+                    .field(TIME_WINDOW_SECONDS_JSON_PROPERTY, listener.time_window_seconds)
+                    .field(BLOCK_EXPIRY_JSON_PROPERTY, listener.block_expiry_seconds)
+                    .field(MAX_BLOCKED_CLIENTS_JSON_PROPERTY, listener.max_blocked_clients)
+                    .field(MAX_TRACKED_CLIENTS_JSON_PROPERTY, listener.max_tracked_clients);
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        };
+    }
+
+    private void authFailureConfigApiRequestHandlers(RequestHandler.RequestHandlersBuilder requestHandlersBuilder) {
+
+        requestHandlersBuilder.override(
+            GET,
+            (channel, request, client) -> loadConfiguration(getConfigType(), false, false).valid(configuration -> {
+                final var config = (ConfigV7) configuration.getCEntry(CType.CONFIG.toLCString());
+                ok(channel, authFailureContent(config));
+            }).error((status, toXContent) -> response(channel, status, toXContent))
+        ).override(DELETE, (channel, request, client) -> loadConfiguration(getConfigType(), false, false).valid(configuration -> {
+            ConfigV7 config = (ConfigV7) configuration.getCEntry(CType.CONFIG.toLCString());
+
+            String listenerName = request.param("name");
+
+            // Try to remove the listener by name
+            if (config.dynamic.auth_failure_listeners.getListeners().remove(listenerName) == null) {
```

Comment:
Is there a potential for an NPE here or will `config.dynamic.auth_failure_listeners` always be non-null?

Reply:
When I was reading the code I thought it would always be initialized based on this [code ref](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/impl/v7/ConfigV7.java#L137). But I am not sure if there are other ways that this would be initialized, can you let me know if this is the correct understanding?

Reply:
You're right. Based on the code-path `config.dynamic.auth_failure_listeners` can never be null`. Resolving this thread.